File: 0098. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#左小右大，直接取該樹的最小值
class Solution(object):
    def __init__(self):
        self.pre = None #用來記錄前一個節點

    def isValidBST(self, root):
        
        if root is None:
            return True

        left = self.isValidBST(root.left)

        if self.pre is not None and self.pre.val >= root.val:
            return False
        self.pre = root

        right = self.isValidBST(root.right)
        return left and right

File: Java/7. Binary Tree/0098. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private long prev = Long.MIN_VALUE; // 初始設定 prev 為最小值，方便比較

    public boolean isValidBST(TreeNode root) {
        if (root == null) return true; // 到達空節點，視為合法

        // 遞迴檢查左子樹，若不合法則直接返回 false
        if (!isValidBST(root.left)) return false;

        // 比較當前節點的值是否大於前一個節點的值
        if (root.val <= prev) return false;

        // 更新 prev 為當前節點的值
        prev = root.val;

        // 遞迴檢查右子樹，若不合法則直接返回 false
        return isValidBST(root.right);
    }
}
